proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,303 @@
+                    self.requests.insert(urn.clone(), next.clone().into());
+                    Ok(next)
+                }
+                None => Err(Error::StateMismatch),
+            },
+        }
+    }
+
+    pub fn requested(&mut self, urn: &RadUrn, timestamp: T) -> Result<Request<Requested, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Created(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.request(timestamp)),
+            urn,
+        )
+    }
+
+    pub fn attempted(
+        &mut self,
+        urn: &RadUrn,
+        attempt: Attempt,
+    ) -> Result<Request<Requested, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.attempt(attempt)),
+            urn,
+        )
+    }
+
+    pub fn found_peer(
+        &mut self,
+        urn: &RadUrn,
+        peer: PeerId,
+        timestamp: T,
+    ) -> Result<Request<Requested, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.found_peer(peer, timestamp)),
+            urn,
+        )
+    }
+
+    pub fn cloning(
+        &mut self,
+        urn: &RadUrn,
+        peer: PeerId,
+        timestamp: T,
+    ) -> Result<Request<Requested, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request),
+                _ => None,
+            },
+            |previous| Ok(previous.cloning(peer, timestamp)),
+            urn,
+        )
+    }
+
+    pub fn fulfilled(
+        &mut self,
+        urn: &RadUrn,
+        timestamp: T,
+        repo: Repo,
+    ) -> Result<Request<Fulfilled, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request),
+                _ => None,
+            },
+            |previous| previous.fulfilled(repo, timestamp),
+            urn,
+        )
+    }
+
+    pub fn canceled(&mut self, urn: &RadUrn, timestamp: T) -> Result<Request<Canceled, T>, Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request),
+                _ => None,
+            },
+            |prev| Ok(prev.cancel(timestamp)),
+            urn,
+        )
+    }
+
+    pub fn timed_out(&mut self, urn: &RadUrn, timestamp: T) -> Result<Request<TimedOut, T>, Error>
```

Comment:
Doesn't seem right that someone is calling this, the waiting room should know what requests are timed out based on the heuristics of `MAX_QUERIES`/`MAX_CLONES`.